File: Coursera/sixteenDiagonals.js
const BLANK = "•";
const SLASH = "/";
const BACKSLASH = "\\";

const FIND_FIRST_SOLUTION = false;
const SOLUTIONS = [];


const copyPermutation = original => {
    const copy = [];
    original.forEach(element => copy.push(element.slice()));
    return copy;
};


const renderBoard = permutation => {
    console.log(permutation.reduce((board, row) => board + row.join("") + "\n", ""));
};


const countDiagonals = permutation => {
    let numberDiagonals = 0;
    permutation.forEach(row => {
        row.forEach(square => {
            if (square !== BLANK) {
                numberDiagonals++;
            }
        });
    });
    return numberDiagonals;
};


const isViable = (permutation, size) => {
    const row = permutation.length - 1;
    if (row > -1) {
        const col = permutation[row].length - 1;
        if (col > -1) {
            if (permutation[row][col] === SLASH) {
                if (col) {
                    if (permutation[row][col - 1] === BACKSLASH) return false;
                }
                if (row) {
                    if (permutation[row - 1][col] === BACKSLASH) return false;
                    if (permutation[row - 1][col + 1] === SLASH) return false;
                }
            }
            if (permutation[row][col] === BACKSLASH) {
                if (col) {
                    if (permutation[row][col - 1] === SLASH) return false;
                    if (row) {
                        if (permutation[row - 1][col - 1] === BACKSLASH) return false;
                    }
                }
                if (row) {
                    if (permutation[row - 1][col] === SLASH) return false;
                }
            }
        }
    }
    return true;
};


const generatePermutations = (permutation, size, diagonals) => {
    if (permutation[permutation.length - 1].length === size) {
        permutation.push([]);
    }
    [SLASH, BACKSLASH, BLANK].forEach(symbol => {
        const newPermutation = copyPermutation(permutation);
        newPermutation[newPermutation.length - 1].push(symbol);
        solve(newPermutation, size, diagonals);
    });
};


const solve = (permutation, size, diagonals) => {
    if (isViable(permutation, size)) {
        if ((permutation.length === size) && (permutation[permutation.length - 1].length === size)) {
            const placedDiagonals = countDiagonals(permutation);
            // const placedDiagonals = permutation.flat().filter(square => square !== BLANK).length; // 7x slower
            if (placedDiagonals === diagonals) {
                if (FIND_FIRST_SOLUTION) {
                    renderBoard(permutation);
                    console.timeEnd("runtime");
                    process.exit(0);
                } else {
                    SOLUTIONS.push(copyPermutation(permutation));
                }
            }
        } else {
            generatePermutations(permutation, size, diagonals);
        }
    }
};


console.time("runtime");
// solve([[]], 3, 6);  // 6.7ms 28 solutions
// solve([[]], 4, 10);  // 82ms 108 solutions
solve([[]], 5, 16);  // 354ms first solution; 29.5s both solutions
console.timeEnd("runtime");
SOLUTIONS.forEach(solution => renderBoard(solution));
console.log(SOLUTIONS.length, "solutions");

File: Coursera/sixteen_diagonals.py
import copy
import time

BLANK = '•'
SLASH = '/'
BACKSLASH = '\\'

FIND_FIRST_SOLUTION = False
SOLUTIONS = []


def render_board(permutation):
    board = ''
    for row in permutation:
        for col in row:
            board += col
        board += "\n"
    print(board)


def is_viable(permutation, size):
    row = len(permutation) - 1
    if row > -1:
        col = len(permutation[row]) - 1
        if col > -1:
            if permutation[row][col] == SLASH:
                if col:
                    if permutation[row][col - 1] == BACKSLASH:
                        return False
                if row:
                    if permutation[row - 1][col] == BACKSLASH:
                        return False
                    try:
                        if permutation[row - 1][col + 1] == SLASH:
                            return False
                    except IndexError:
                        pass
            if permutation[row][col] == BACKSLASH:
                if col:
                    if permutation[row][col - 1] == SLASH:
                        return False
                    if row:
                        if permutation[row - 1][col - 1] == BACKSLASH:
                            return False
                if row:
                    if permutation[row - 1][col] == SLASH:
                        return False
    return True


def solve(permutation, size, diagonals):
    if is_viable(permutation, size):
        if len(permutation) == size and len(permutation[-1]) == size:
            placed_diagonals = 0
            for row in permutation:
                placed_diagonals += size - row.count(BLANK)
            if placed_diagonals == diagonals:
                if FIND_FIRST_SOLUTION:
                    render_board(permutation)
                    global start
                    runtime = time.perf_counter() - start
                    print(f"{runtime:0.4f} seconds")
                    exit()
                else:
                    global SOLUTIONS
                    SOLUTIONS.append(copy.deepcopy(permutation))
        else:
            generate_permutations(permutation, size, diagonals)


def generate_permutations(permutation, size, diagonals):
    if len(permutation[-1]) == size:
        permutation.append([])
    for symbol in (SLASH, BACKSLASH, BLANK):
        new_permutation = copy.deepcopy(permutation)
        new_permutation[-1].append(symbol)
        solve(new_permutation, size, diagonals)


start = time.perf_counter()
# solve([[]], 3, 6)  # 350ms
# solve([[]], 4, 10)  # 6s
solve([[]], 5, 16)  # 30s first solution, 53m both solutions
runtime = time.perf_counter() - start
for solution in SOLUTIONS:
    render_board(solution)
print(len(SOLUTIONS), f"solutions in {runtime:0.3f} seconds")
